invenio_app_ils/holidays/api.py

```diff
@@ -0,0 +1,84 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2020 CERN.
+#
+# invenio-app-ils is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""ILS holidays API."""
+
+from datetime import timedelta
+
+import arrow
+
+from invenio_app_ils.proxies import current_app_ils
+
+_day_increment = timedelta(days=1)  # Atomic increment
+
+
+def _is_normally_open(location, date):
+    """Checks if the location is normally opened on a given date wrt to the regular schedule."""
+    opening = location["opening_weekdays"]
+    weekday_number = date.weekday()
+    opening_weekday = opening[weekday_number]
+    return opening_weekday["is_open"]
+
+
+def _is_exceptionally_open(location, date):
+    """Checks if the location is exceptionally opened on a given date, or None if undefined."""
+    for exception in location["opening_exceptions"]:
+        if date <= arrow.get(exception["end_date"]).date():
+            if arrow.get(exception["start_date"]).date() <= date:
+                return exception["is_open"]
+            else:
+                return None  # Early return
+    return None  # Exhaustion
+
+
+def is_open_on(location_pid, date):
+    """Checks if the location is open or closed on a given date."""
+    location = current_app_ils.location_record_cls.get_record_by_pid(location_pid)
+    exceptionally_open = _is_exceptionally_open(location, date)
+    if exceptionally_open is not None:
+        return exceptionally_open
+    return _is_normally_open(location, date)
+
+
+def next_open_after(location_pid, date):
```

Comment:
I am sure this is the first attempt and needs refinement, if it can help, here my pseudocode:

```
def _date_in_range(date, location_exceptions):
    for ex in location_exceptions:
        if ex.start <= TODAY_CHECK <= ex.end:
            return True
    return False


def is_working_day(date, location_weekdays, location_exceptions_is_open, location_exceptions_is_closed):
  day_name = date.get_day_name
  weekday = location_weekdays[day_name]
  if weekday.is_open and not _date_in_range(date, location_exceptions_is_closed):
    return True
  if weekday.is_closed and _date_in_range(date, location_exceptions_is_open):
    return True
  return False
  

def _get_next_working_day(date, weekdays, exceptions_is_open, exceptions_is_closed):
    if is_working_day(date, weekdays, exceptions_is_open, exceptions_is_closed):
      return date
    else:
      return _get_next_working_day(date + timedelta(days=1), weekdays, exceptions_is_open, exceptions_is_closed)
  

def get_next_working_day(location, date):
  weekdays, exceptions_is_open, exceptions_is_closed = get_...(location)
  return _get_next_working_day(date, weekdays, exceptions_is_open, exceptions_is_closed)
```

The idea is:
```
while True (because at some point there must be a date available, otherwise the library is closed!)
  1. if date is normally open, then check if it is exceptionally closed
  2. if date is normally closed, then check if it is exceptionally open 
 date = date+1 day
```

Reply:
Thanks for that!
It's more or less what RERO does, this procedure I wrote is essentially a more efficient implementation of that although I agree that will not make much of a difference. One must also take into account the special case where all the weekdays are being closed and the library only relies on opening exceptions (this is already handled in the current procedure).